### ISSUE TITLE
TPOT Output Inconsistent with Exported Pipeline Crossval Results

### ISSUE THREAD

**Issue by CSNoyes** [rhiever/tpot#513]
Using TPOT to optimize a classifier on a set of time-series data, using any of the exported pipelines (with identical data, cross-validator, model) gives different (and significantly worse) results.

To reproduce, I've uploaded the dataset here: https://drive.google.com/file/d/0B_shzyNv2gBQdGdHTHNTUEdsU3M/view?usp=sharing
```
import pandas as pd
import numpy as np
from sklearn.model_selection import TimeSeriesSplit
from tpot import TPOTClassifier

data = pd.read_csv("data.csv", index_col=False)
features = data.drop("target", axis=1).as_matrix()[:,1:]
targets = np.array(data["target"])
cv = TimeSeriesSplit(n_splits=30)

tpot = TPOTClassifier(generations=10, population_size=100, offspring_size=100, verbosity=2, cv=cv)
tpot.fit(features, targets)
tpot.export("tpotExportedPipeline.py")
```

Running the below snippet, I get outputs that get to about 60% accuracy, like the following:
```
exported_pipeline = make_pipeline(
    GradientBoostingClassifier(learning_rate=1.0, max_depth=2, max_features=0.3, min_samples_leaf=7, min_samples_split=7, subsample=0.25)
)
```

But when running the pipeline on the exact same features/targets variables and the below cross-val scorer, I get wildly different accuracies. Could anyone offer some insight on why this inconsistency exists?
```
scores = cross_val_score(exported_pipeline, features, targets, cv=cv, n_jobs=-1)
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))
Accuracy: .5 (+/- 1.0)
```

**Comment by weixuanfu2016** [comment]
I think this issue is about non `random_state` in export codes in TPOT output. Below is a small demo to reproduce the issue. You may also use the demo to add `random_state` to your pipelines for consistent CV score. (Note, the old API of xgboost use `seed` for random seeds.)

```
# coding: utf-8
import pandas as pd
import numpy as np
from sklearn.model_selection import TimeSeriesSplit, cross_val_score
from tpot import TPOTClassifier
from tpot.export_utils import generate_pipeline_code, expr_to_tree

data = pd.read_csv("data.csv", index_col=False)
features = data.drop("target", axis=1).as_matrix()[:,1:]
targets = np.array(data["target"])
cv = TimeSeriesSplit(n_splits=30)

tpot = TPOTClassifier(generations=2, population_size=10, offspring_size=10, verbosity=2, cv=cv, random_state=42)
tpot.fit(features, targets)
# print out pipeline
print(tpot.fitted_pipeline_)
sklearn_pipeline_str = generate_pipeline_code(expr_to_tree(tpot._optimized_pipeline, tpot._pset), tpot.operators)
print(sklearn_pipeline_str)
sklearn_pipeline = eval(sklearn_pipeline_str, tpot.operators_context)
print(sklearn_pipeline) # should be the same as tpot.fitted_pipeline_
scores = cross_val_score(sklearn_pipeline, features, targets, cv=cv, n_jobs=-1)
print("Accuracy without random_state:", scores.mean())
tpot._set_param_recursive(sklearn_pipeline.steps, 'random_state', 42)
if 'XGB' in sklearn_pipeline_str: # for old API in XGBClassifier 
    tpot._set_param_recursive(sklearn_pipeline.steps, 'seed', 42)
scores = cross_val_score(sklearn_pipeline, features, targets, cv=cv, n_jobs=-1)
print("Accuracy with random_state:", scores.mean())
```

**Comment by CSNoyes** [comment]
Fantastic, thank you. A suggestion: for problems where the data is extremely noisy, it might be good to make TPOT reseed the random weights of each model at each generation, such that a model that is good in a very noisy dataset with a certain random weight will be less likely to be selected as a top performing genome.
